source/common/upstream/cluster_manager_impl.cc

```diff
@@ -844,27 +844,35 @@ ThreadLocalCluster* ClusterManagerImpl::getThreadLocalCluster(absl::string_view
   }
 }
 
-void ClusterManagerImpl::maybePrefetch(
+void ClusterManagerImpl::maybePreconnect(
     ThreadLocalClusterManagerImpl::ClusterEntry& cluster_entry,
-    std::function<ConnectionPool::Instance*()> pick_prefetch_pool) {
-  // TODO(alyssawilk) As currently implemented, this will always just prefetch
-  // one connection ahead of actually needed connections.
-  //
-  // Instead we want to track the following metrics across the entire connection
-  // pool and use the same algorithm we do for per-upstream prefetch:
-  // ((pending_streams_ + num_active_streams_) * global_prefetch_ratio >
-  //  (connecting_stream_capacity_ + num_active_streams_)))
-  //  and allow multiple prefetches per pick.
-  //  Also cap prefetches such that
-  //  num_unused_prefetch < num hosts
-  //  since if we have more prefetches than hosts, we should consider kicking into
-  //  per-upstream prefetch.
-  //
-  //  Once we do this, this should loop capped number of times while shouldPrefetch is true.
-  if (cluster_entry.cluster_info_->peekaheadRatio() > 1.0) {
-    ConnectionPool::Instance* prefetch_pool = pick_prefetch_pool();
-    if (prefetch_pool) {
-      prefetch_pool->maybePrefetch(cluster_entry.cluster_info_->peekaheadRatio());
+    const ClusterConnectivityState& state,
+    std::function<ConnectionPool::Instance*()> pick_preconnect_pool) {
+  auto peekahead_ratio = cluster_entry.cluster_info_->peekaheadRatio();
+  if (peekahead_ratio <= 1.0) {
+    return;
+  }
+
+  // 3 here is arbitrary. Just as in ConnPoolImplBase::tryCreateNewConnections
+  // we want to limit the work which can be done on any given preconnect attempt.
+  for (int i = 0; i < 3; ++i) {
+    // Just as in ConnPoolImplBase::shouldCreateNewConnection, see if adding this one new connection
+    // would put the cluster over desired capacity. If so, stop preconnecting.
+    if ((state.pending_streams_ + 1 + state.active_streams_) * peekahead_ratio <=
+        (state.connecting_stream_capacity_ + state.active_streams_)) {
+      return;
+    }
+    ConnectionPool::Instance* preconnect_pool = pick_preconnect_pool();
+    if (preconnect_pool) {
```

Comment:
nit: I think this can be `if (!preconnect_pool || !preconnect_pool->maybePreconnect(peekahead_ratio)) { return; }` but I might have gotten that wrong.